File: src/main/java/app/CipherEngine.java
package app;

import javax.crypto.Cipher;
import javax.crypto.SecretKeyFactory;
import javax.crypto.spec.GCMParameterSpec;
import javax.crypto.spec.PBEKeySpec;
import javax.crypto.spec.SecretKeySpec;
import java.nio.ByteBuffer;
import java.nio.charset.StandardCharsets;
import java.security.SecureRandom;
import java.util.Arrays;
import java.util.Base64;

public final class CipherEngine {
    private static final byte[] MAGIC = "CJFX1".getBytes(StandardCharsets.US_ASCII);
    private static final byte[] LEGACY_SALT = "token-crypt:v1:fixed-salt".getBytes(StandardCharsets.UTF_8);

    private static final int SALT_BYTES = 16;
    private static final int ITERATIONS = 300_000;
    private static final int KEY_BITS = 256;
    private static final int GCM_TAG_BITS = 128;
    private static final int GCM_TAG_BYTES = GCM_TAG_BITS / 8;
    private static final int IV_BYTES = 12;

    private static final SecureRandom RNG = new SecureRandom();

    private CipherEngine() {}

    private static SecretKeySpec keyFromToken(String token, byte[] salt) throws Exception {
        char[] password = token.toCharArray();
        byte[] keyBytes = null;
        PBEKeySpec spec = null;

        try {
            spec = new PBEKeySpec(password, salt, ITERATIONS, KEY_BITS);
            SecretKeyFactory f = SecretKeyFactory.getInstance("PBKDF2WithHmacSHA256");
            keyBytes = f.generateSecret(spec).getEncoded();
            return new SecretKeySpec(keyBytes, "AES");
        } finally {
            if (spec != null) spec.clearPassword();
            Arrays.fill(password, '\0');
            if (keyBytes != null) Arrays.fill(keyBytes, (byte) 0);
        }
    }

    public static String encrypt(String token, String plaintext) throws Exception {
        byte[] salt = new byte[SALT_BYTES];
        byte[] iv = new byte[IV_BYTES];

        RNG.nextBytes(salt);
        RNG.nextBytes(iv);

        SecretKeySpec key = keyFromToken(token, salt);

        Cipher cipher = Cipher.getInstance("AES/GCM/NoPadding");
        cipher.init(Cipher.ENCRYPT_MODE, key, new GCMParameterSpec(GCM_TAG_BITS, iv));

        byte[] pt = plaintext == null ? new byte[0] : plaintext.getBytes(StandardCharsets.UTF_8);
        byte[] ct = cipher.doFinal(pt);

        // output = base64( magic/version || salt || iv || ciphertext+tag )
        ByteBuffer bb = ByteBuffer.allocate(MAGIC.length + salt.length + iv.length + ct.length);
        bb.put(MAGIC);
        bb.put(salt);
        bb.put(iv);
        bb.put(ct);

        return Base64.getEncoder().encodeToString(bb.array());
    }

    public static String decrypt(String token, String tokenText) throws Exception {
        if (tokenText == null || tokenText.isBlank()) {
            throw new IllegalArgumentException("Ciphertext is empty");
        }

        byte[] all = Base64.getDecoder().decode(tokenText.trim());

        if (isCurrentFormat(all)) {
            return decryptCurrent(token, all);
        }

        return decryptLegacy(token, all);
    }

    private static boolean isCurrentFormat(byte[] all) {
        if (all.length < MAGIC.length + SALT_BYTES + IV_BYTES + GCM_TAG_BYTES) return false;

        for (int i = 0; i < MAGIC.length; i++) {
            if (all[i] != MAGIC[i]) return false;
        }

        return true;
    }

    private static String decryptCurrent(String token, byte[] all) throws Exception {
        ByteBuffer bb = ByteBuffer.wrap(all);
        bb.position(MAGIC.length);

        byte[] salt = new byte[SALT_BYTES];
        byte[] iv = new byte[IV_BYTES];
        byte[] ct = new byte[bb.remaining() - SALT_BYTES - IV_BYTES];

        bb.get(salt);
        bb.get(iv);
        bb.get(ct);

        return decryptBytes(token, salt, iv, ct);
    }

    private static String decryptLegacy(String token, byte[] all) throws Exception {
        if (all.length < IV_BYTES + GCM_TAG_BYTES) {
            throw new IllegalArgumentException("Ciphertext too short");
        }

        byte[] iv = new byte[IV_BYTES];
        byte[] ct = new byte[all.length - IV_BYTES];

        System.arraycopy(all, 0, iv, 0, IV_BYTES);
        System.arraycopy(all, IV_BYTES, ct, 0, ct.length);

        return decryptBytes(token, LEGACY_SALT, iv, ct);
    }

    private static String decryptBytes(String token, byte[] salt, byte[] iv, byte[] ct) throws Exception {
        SecretKeySpec key = keyFromToken(token, salt);

        Cipher cipher = Cipher.getInstance("AES/GCM/NoPadding");
        cipher.init(Cipher.DECRYPT_MODE, key, new GCMParameterSpec(GCM_TAG_BITS, iv));

        byte[] pt = cipher.doFinal(ct);
        return new String(pt, StandardCharsets.UTF_8);
    }
}

File: lib/cipher_engine.dart
import 'dart:convert';
import 'dart:math';
import 'dart:typed_data';

import 'package:cryptography/cryptography.dart';

final class CipherEngine {
  static const List<int> _magic = <int>[67, 74, 70, 88, 49]; // CJFX1
  static const List<int> _legacySalt = <int>[
    116, 111, 107, 101, 110, 45, 99, 114, 121, 112, 116, 58, 118, 49, 58,
    102, 105, 120, 101, 100, 45, 115, 97, 108, 116,
  ];

  static const int _saltBytes = 16;
  static const int _iterations = 300000;
  static const int _keyBits = 256;
  static const int _tagBytes = 16;
  static const int _ivBytes = 12;

  static final Random _rng = Random.secure();
  static final Pbkdf2 _pbkdf2 = Pbkdf2(
    macAlgorithm: Hmac.sha256(),
    iterations: _iterations,
    bits: _keyBits,
  );
  static final AesGcm _aesGcm = AesGcm.with256bits(nonceLength: _ivBytes);

  CipherEngine._();

  static Future<String> encrypt(String token, String plaintext) async {
    final salt = _randomBytes(_saltBytes);
    final iv = _randomBytes(_ivBytes);
    final key = await _keyFromToken(token, salt);

    final box = await _aesGcm.encrypt(
      utf8.encode(plaintext),
      secretKey: key,
      nonce: iv,
    );

    final payload = <int>[
      ..._magic,
      ...salt,
      ...iv,
      ...box.cipherText,
      ...box.mac.bytes,
    ];

    return base64.encode(payload);
  }

  static Future<String> decrypt(String token, String tokenText) async {
    final trimmed = tokenText.trim();
    if (trimmed.isEmpty) {
      throw ArgumentError('Ciphertext is empty');
    }

    final all = base64.decode(trimmed);
    if (_isCurrentFormat(all)) {
      return _decryptCurrent(token, all);
    }
    return _decryptLegacy(token, all);
  }

  static Future<SecretKey> _keyFromToken(String token, List<int> salt) {
    return _pbkdf2.deriveKeyFromPassword(password: token, nonce: salt);
  }

  static bool _isCurrentFormat(List<int> all) {
    if (all.length < _magic.length + _saltBytes + _ivBytes + _tagBytes) {
      return false;
    }
    for (var i = 0; i < _magic.length; i++) {
      if (all[i] != _magic[i]) return false;
    }
    return true;
  }

  static Future<String> _decryptCurrent(String token, List<int> all) {
    var offset = _magic.length;
    final salt = all.sublist(offset, offset + _saltBytes);
    offset += _saltBytes;
    final iv = all.sublist(offset, offset + _ivBytes);
    offset += _ivBytes;
    final encryptedAndTag = all.sublist(offset);

    return _decryptBytes(token, salt, iv, encryptedAndTag);
  }

  static Future<String> _decryptLegacy(String token, List<int> all) {
    if (all.length < _ivBytes + _tagBytes) {
      throw ArgumentError('Ciphertext too short');
    }

    final iv = all.sublist(0, _ivBytes);
    final encryptedAndTag = all.sublist(_ivBytes);
    return _decryptBytes(token, _legacySalt, iv, encryptedAndTag);
  }

  static Future<String> _decryptBytes(
    String token,
    List<int> salt,
    List<int> iv,
    List<int> encryptedAndTag,
  ) async {
    if (encryptedAndTag.length < _tagBytes) {
      throw ArgumentError('Ciphertext too short');
    }

    final split = encryptedAndTag.length - _tagBytes;
    final cipherText = encryptedAndTag.sublist(0, split);
    final macBytes = encryptedAndTag.sublist(split);
    final key = await _keyFromToken(token, salt);

    final clearBytes = await _aesGcm.decrypt(
      SecretBox(cipherText, nonce: iv, mac: Mac(macBytes)),
      secretKey: key,
    );

    return utf8.decode(clearBytes);
  }

  static Uint8List _randomBytes(int length) {
    return Uint8List.fromList(
      List<int>.generate(length, (_) => _rng.nextInt(256)),
    );
  }
}
